Show error page when feed fails to load fix #4024

fffw opened the pull request:
@myleshorton I made a few changes based on your work.

fffw commented on src/github.com/getlantern/lantern-ui/app/partials/error.html:
```diff
@@ -0,0 +1,46 @@
+  </div>
+  <div id="error-reasons">
+    <ol>
+      <li ng-show="!model.settings.systemProxy">
```
Related to https://github.com/getlantern/lantern/issues/4022

myleshorton replied:
I don't think we want this @fffw, as this is a very common issue with the system proxy on or off -- i.e. some issue with extensions right?

myleshorton replied:
Oh hey sorry I misread this @fffw -- looks good!

fffw commented on src/github.com/getlantern/lantern-ui/app/partials/error.html:
```diff
@@ -0,0 +1,46 @@
+        <div ng-show="!showConnectionHelp"><a ng-click="toggleShowConnectionHelp()">{{ 'WHAT_DO' | translate }}</a></div>
+        <div class="error-action ng-hide" ng-show="showConnectionHelp">
+          <div>{{ 'CONNECTION_HELP' | translate }}</div>
+          <div>Windows Vista/7/8/10: C:\Users\USERNAME\AppData\Roaming\Lantern\logs</div>
```
It seems too detailed to show path names on this page. Can we just add the link to https://github.com/getlantern/lantern/wiki/Emailing-lantern-logs?

myleshorton replied:
Yeah that was the original design. I could go either way @fffw. The other thing I was playing with was actually including the links in the translated version but didn't quite get it working.

fffw replied:
@myleshorton I don't quite understand. Do you mean to set up different link for each translation?

myleshorton replied:
No, just that we need "Facebook", "GitHub", and "Forums" to actually link to those places, but we're linking from translated text.

fffw replied:
Ok I see. I can do it the same way we share mobile download link with different languages.

myleshorton replied:
Honestly I think there are also cleaner ways to do it with angular translate but maybe I'm wrong. I actually thought your `$sce.trustAsHtml` from services.js might do the trick no?

fffw commented on src/github.com/getlantern/lantern-ui/app/angular-feed/scripts/services.js:
```diff
@@ -37,7 +37,8 @@ angular.module('feeds-services', []).factory('feedService', ['$q', '$http', '$sc
         }
-        feedCache.set(feedURL, data);
+        // Always get fresh copy, in hope to be more reactive to network errors
+        // feedCache.set(feedURL, data);
```
It's still not unclear to me if we should keep the 5 minutes caching or not. Occasionally it took me around 30 seconds to get a response, even when the response is "304 Not modified". Caching would help for frequent refreshes.

myleshorton replied:
Sounds like maybe we should keep it then? Very odd it could sometimes take so long. Is your proxy generally slow at times when you've seen it take a long time to load the feed?

fffw replied:
It doesn't happen to me now, could be caused by network glitches. It may take a bit longer to load feeds when Lantern starts up and old config doesn't work any more.

One possible way is to load cached feeds first, fetch feeds in background and replace it when succeeded, or show error page if failed. Not sure if I have a chance to make that change today.

oxtoacart replied:
+1 on caching on the client

cc @xiam

myleshorton commented:
Yeah aside from the links in the translations @fffw this seems good to go to me. The only thing I haven't been able to verify is that the on-error handler actually gets called when there's an error. When I just changed it to some bogus URL, it actually did not get called, but maybe there's a better way. Have you seen the on-error handler actually get called?

myleshorton commented:
OK @fffw I just added the links and also threw in OS-detection so we can tell the user exactly where their logs are. Also removed d3 while in there.

myleshorton commented:
Hey @fffw I know you're on break, but you want to give this a look when you're back? Thanks!

fffw commented:
@myleshorton Device OS detection is a cool idea! 

I usually trigger error page by turning off Wifi temporarily. When changing to bogus URL, both `feedUrl()` and `fallback-url` need to be changed.

I added the logic to show cached feeds first if any, and fetch server in the background. It will re-render feeds after a successful fetching, or render the error page. I also adjusted the cache from 5 minutes to 24 hours, so users don't need to wait for the feeds to load in most cases.

myleshorton commented:
This all looks good to me @fffw! Merging.

fffw commented:
Thanks @myleshorton !
